cEP-0018.md

```diff
@@ -0,0 +1,276 @@
+          ext = Languages(parse_lang_str(lang)[0])[0]
+        if ext is None:
+            raise FileExtensionNotSupported('Unknown File Type')
+        inputStream = file.readlines()
+        lexer = mapping[ext][0](inputStream)
+        parser = mapping[ext][1](lexer)
+        return parser
+
+class ASTWalker():
+    treeRoot = None
+    tree = None
+
+    @staticmethod
+    def make_walker():
+        """
+        Resolves walker object for appropriate type
+        """
+        if lang == 'auto':
+            return ASTWalker
+        else:
+            return resolve_with_package_info(lang)
+
+    @staticmethod
+    def get_walker(file, filename, lang):
+        """
+        Returns a walker to the required language
+        """
+        ret_class = make_walker(lang)
+        if not ret_class:
+            raise Exception('Un-supported Language')
+        else:
+            return ret_class(file, filename)
+
+    def __init__(file,filename):
+        self.tree = ASTLoader.load(file, filename)
+        self.treeRoot = self.tree
+
+    def get_next_node(self):
+        """
+        Modify the internal tree variable and return the next node
+        """
+
+    def get_prev_node(self):
+        """
+        Modify the internal tree variable and return prev node
+        """
+
+    def reset(self):
+        """
+        Reset the internal tree variable to tree root
+        """
+
+    def is_terminal_node(node):
+        """
+        Determine if the node is terminal
+        """
+        return isinstance(node, antlr4.tree.Tree.TerminalNode)
+    ...
+```
+
+### Prototype of Py3ASTWalker
+
+These kinds of Walkers will supply a high level walker api
+
+```python
+class Py3ASTWalker(ASTWalker):
+    def next_function():
+        # modify the tree node variable such that it only stops at functions
+        # which can be detected by the Python grammar rules
+    ...
+```
+
+### Prototype of  `ASTBear` class implementation
+
+```python
+from coantlib import ASTWalker
+from coalib.bears.LocalBear import Bear
+
+class ASTBear(LocalBear):
+    walker = None
+
+    def initialise(file, filename):
+        if LANGUAGES:
+            self.walker = ASTWalker.get_walker(file,
+                                               filename,
+                                               self.LANGUAGES[0])
+        else:
+            self.walker = ASTWalker.get_walker(file, filename, 'auto')
+
+    def run(self,
+            filename,
+            file,
+            tree,
+            *args,
+            dependency_results = None,
+            **kwargs
+            ):
+        raise NotImplementedError('Needs to be done by bear')
+```
+
+### Prototype of  `Py3ASTBear` class implementation
+
+These kinds of base bears will provide an even more abstracted high level api
+Following is a demonstration in case language grammars have common things in them
+(For e.g Python2 and Python3)
+
+```python
+from coantlib import ASTBear
+
+class BasePyASTBear(ASTBear):
```

Comment:
What is the purpose of this bear. Does it provide any intermediate results? If yes what results does it provide?